Add extensions api

ravitemer opened the pull request:
## Description

As discussion in #1105, having an extensions API makes it easy to extend and share codecompanion. It is something similar to telescope extensions. It allows something like below

## Using Extensions

Extensions are configured in your CodeCompanion setup:

```lua
-- Install the extension
{
  "olimorris/codecompanion.nvim",
  dependencies = {
    "author/cc_history.nvim" -- Adds autosaving chats, keymaps etc
  }
}

-- Configure in your setup
require("codecompanion").setup({
  extensions = {
    cc_history = {
      enabled = true, -- Optional, defaults to true
      opts = {
        -- Extension specific options
      }
    }
  }
})
```

Once configured, extension exports are accessible via:

```lua
local codecompanion = require("codecompanion")
-- Use exported functions
codecompanion.extensions.cc_history.some_function()
```


 
<!-- Describe the big picture of your changes to communicate to the maintainers why we should accept this pull request. -->

## Related Issue(s)

<!--
  If this PR fixes any issues, please link to the issue here.
  - Fixes #<issue_number>
-->

## Screenshots

<!-- Add screenshots of the changes if applicable. -->

## Checklist

- [x] I've read the [contributing](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines and have adhered to them in this PR
- [x] I've updated the README and/or relevant docs pages
- [x] I've run `make docs` to update the vimdoc pages

ravitemer commented:
I think we could also have a page like "Extensions" where users can open PRs to submit their extensions with title, brief description , installation instructions.

olimorris commented:
This is amazing. Super, super exciting! And appreciate you sticking to my coding patterns as well.

ravitemer commented:
I feel it's best if you merge this PR along with the function tools migration all with the next v15. I have upgraded the mcphub extension to use this API.  

### Old Way

```lua
require("mcphub").setup({
    extensions = {
        codecompanion = {
            -- Show the mcp tool result in the chat buffer
            show_result_in_chat = true,
            make_vars = true, -- make chat #variables from MCP server resources
            make_slash_commands = true, -- make /slash_commands from MCP server prompts
        },
    }
})
```
```lua
require("codecompanion").setup({
    strategies = {
        chat = {
            tools = {
                ["mcp"] = {
                    -- calling it in a function would prevent mcphub from being loaded before it's needed
                    callback = function() return require("mcphub.extensions.codecompanion") end,
                    description = "Call tools and resources from the MCP Servers",
                }
            }
        }
    }
})
```

### Current
```lua
require("codecompanion").setup({
--other config
extensions = {
  mcphub = {
    callback = "mcphub.extensions.codecompanion",
    opts = {
      make_vars = true,
      make_slash_commands = true,
      show_result_in_chat = true,
    },
  },
 },
})
```


@Davidyz I think vectorcode can also be modified to use the API. This way all the changes that related to v15 along with mcphub and vectorcode can be documented in the announcement discussion. 

We can add the new Extensions page in docs website starting with vectorcode and mcphub so that the community can build upon that.

Davidyz commented:
So basically this wraps the slash commands and tools into one interface for the users when setting up codecompanion? I feel like this might be an overkill for vectorcode because the Vectorcode tool and slash command basically do the same thing, and users usually need only one of them. That said, I can see how this is more convenient for MCP and I'm excited to try this out if it gets merged!

ravitemer commented:
Yes! basically add any keymaps, slash_commands and tools in a setup() function. If it is standalone codecompaion extension it resolves the extension from `{reponame}.codecompanion._extensions.{extension_name}`. If it is not a plugin just for codecompanion like mcphub and vectorcode we specify the path in callback field that exports a table with a setup(opts) function where opts are the opts given in codecompanion.config.extensions.extension_name.opts

I feel using extensions API makes it resilient for any version or API changes in the future where the user doesn't need to do any changes.

Davidyz commented:
That sounds cool. Potentially, we could just omit the `extensions` field in codecompanion setup opts, and just make it conventional for the extension creators to call the extension manager `resolve` (or is it `load_extension`?) method to register the extension. The extensions (vectorcode, mcphub, etc.) can then have their own setup options for the users to decide whether to register the extension automatically, similar to how codecompanion is now handling [blink and nvim-cmp](https://github.com/olimorris/codecompanion.nvim/blob/a382a8ced48c7664bd3dad992ec9f9b8054dc952/plugin/codecompanion.lua#L67). This would further minimise the modifications to the codecompanion `setup`.

Also, since this is meant to be a wrapper around the existing tool/slash command system, it'd be nice to include some brief instructions on how to register tools/slash commands from within an extension.

ravitemer commented:
> That sounds cool. Potentially, we could just omit the `extensions` field in codecompanion setup opts, and just make it conventional for the extension creators to call the extension manager `resolve` (or is it `load_extension`?) method to register the extension. The extensions (vectorcode, mcphub, etc.) can then have their own setup options for the users to decide whether to register the extension automatically, similar to how codecompanion is now handling [blink and nvim-cmp](https://github.com/olimorris/codecompanion.nvim/blob/a382a8ced48c7664bd3dad992ec9f9b8054dc952/plugin/codecompanion.lua#L67). This would further minimise the modifications to the codecompanion `setup`.
> 
> Also, since this is meant to be a wrapper around the existing tool/slash command system, it'd be nice to include some brief instructions on how to register tools/slash commands from within an extension.

I have added `register_extension` for dynamic registration. 

### Dynamic registration

Extensions can also be added dynamically using 

```lua
require("codecompanion").register_extension("codecompanion_history", {
  callback = {
    setup = function()
      -- Add action to chat keymaps
      local chat_keymaps = require("codecompanion.config").strategies.chat.keymaps

      chat_keymaps.open_saved_chats = {
        modes = {
          n = opts.keymap or "gh",
        },
        description = "Open Saved Chats",
        callback = function(chat)
          -- Implementation of opening saved chats
          vim.notify("Opening saved chats for " .. chat.id)
        end,
      }
    end,
    exports = {},
  },
})
```

ravitemer commented:
> Potentially, we could just omit the extensions field in codecompanion setup opts, and just make it conventional for the extension creators to call the extension manager resolve (or is it load_extension?) method to register the extension.

@Davidyz Yeah! We can also do that. I basically followed telescope way of adding extensions. Whatever @olimorris decides. I personally prefer having the extensions field in the codecompanion config as all that is related to codecompanion stays in one place rather than at individual plugins.

olimorris commented:
> I personally prefer having the extensions field in the codecompanion config as all that is related to codecompanion stays in one place rather than at individual plugins.

I agree. As a user with a sprawling config, if everything related to CodeCompanion can sit within that config and setup, then I won't have to think or search for it.

@ravitemer - I think it's worth putting a small paragraph in the installation docs page. Just as a note for how extensions are installed perhaps even show how you'd get mcphub working. Whilst it's not the responsibility of CodeCompanion to document how all the potential extensions can be installed, I think it would be a nice primer for anyone that's new and comes to the plugin. Could we also add some basic tests as well? I want to make sure something as important as this at least has some coverage.

olimorris commented:
> I feel it's best if you merge this PR along with the function tools migration all with the next v15

And agree on this. I'll wait for you both to give me the nod before I do.

ravitemer commented:
> I want to make sure something as important as this at least has some coverage.

Sure! Wanted to make sure you are okay with the API. Will add tests and update installation docs.

olimorris commented:
Oh and we should also update `codecompanion.has` and include this as a feature. I've updated the PR template and added a checkbox for this.

ravitemer commented:
@olimorris I think this is complete. Added tests, updated docs mainly installation page, configuration section and added "extensions" to has()

In addition I have created a new section in the sidebar called "Community Extensions". I have included a page in that for mcphub as an example detailing the complete integration. 

Please let me know if there is anything missing.

ravitemer commented:
Please add this along with the v15 release announcement

<details>
<summary> mcphub.nvim migration </summary>

[mcphub.nvim](https://github.com/ravitemer/mcphub.nvim) is now available as an extension that can be installed as follows. It works with both xml and function tools. Please visit [codecompanion's mcphub extension](https://codecompanion.olimorris.dev/extensions/mcphub.html) for complete documentation.

1. Remove old mcp tool from `chat.tools`. 

```lua
require("codecompanion").setup({
 -- other codecompanion config
  strategies = {
    chat = {
       tools = {
     -- Remove old mcp tool which will be auto added as a tool group with two individual tools.
      --   ["mcp"] = {
      --     callback = function()
      --       return require("mcphub.extensions.codecompanion")
      --     end,
      --     description = "Call tools and resources from MCP Servers",
      --   },
     },
    },
  },
})
```

2. Add extension

```lua
require("codecompanion").setup({
--other config
extensions = {
  mcphub = {
    callback = "mcphub.extensions.codecompanion",
    opts = {
      make_vars = true,
      make_slash_commands = true,
      show_result_in_chat = true,
    },
  },
 },
})
```
* `mcphub.config.extensions.codecompanion` is deprecated as the options
are now declared at codecompanion extension itself.

</details>

ravitemer commented:
Currently mcphub examples are everywhere in the extensions docs 😅, will change them once a standalone codecompanion extension is available.

Davidyz commented:
Neat! It'll probably easier for me to start the migration after the function-call release though, because I refactored the tools quite heavily in vectorcode's function-calling PR and I don't want to make the git branches too complicated. Thanks for the effort!

olimorris commented:
Wow @ravitemer this is some work! Really, really appreciate your efforts on contributing to the docs as well. You've added a lot of comprehensive examples which will make it super easy for people to leverage and start (hopefully) building their own extensions.

LGTM!

olimorris commented:
I was thinking I'd merge this into `main` first, create a branch called `has_xml_tools` that users who want to stay on the old tool versions can use, then merge `function_calling` after. This would all be done right away.

ravitemer commented:
Perfect!

olimorris commented:
@ravitemer let me know when you're good to go

olimorris commented:
I notice CloudFlare isn't building the doc site out in this PR. Could you run `npm --prefix doc/ run build` on your machine to check it builds okay?

ravitemer commented:
@olimorris I've fixed a dead link. 

Yes! Nothing from my side.
